src/rust/src/wallet_scanner.rs

```diff
@@ -0,0 +1,478 @@
+                consensus::NetworkUpgrade::Sapling => *sapling,
+                consensus::NetworkUpgrade::Blossom => *blossom,
+                consensus::NetworkUpgrade::Heartwood => *heartwood,
+                consensus::NetworkUpgrade::Canopy => *canopy,
+                consensus::NetworkUpgrade::Nu5 => *nu5,
+            },
+        }
+    }
+
+    fn coin_type(&self) -> u32 {
+        match self {
+            Self::Consensus(params) => params.coin_type(),
+            Self::RegTest { .. } => constants::regtest::COIN_TYPE,
+        }
+    }
+
+    fn hrp_sapling_extended_spending_key(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_extended_spending_key(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_SPENDING_KEY,
+        }
+    }
+
+    fn hrp_sapling_extended_full_viewing_key(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_extended_full_viewing_key(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY,
+        }
+    }
+
+    fn hrp_sapling_payment_address(&self) -> &str {
+        match self {
+            Self::Consensus(params) => params.hrp_sapling_payment_address(),
+            Self::RegTest { .. } => constants::regtest::HRP_SAPLING_PAYMENT_ADDRESS,
+        }
+    }
+
+    fn b58_pubkey_address_prefix(&self) -> [u8; 2] {
+        match self {
+            Self::Consensus(params) => params.b58_pubkey_address_prefix(),
+            Self::RegTest { .. } => constants::regtest::B58_PUBKEY_ADDRESS_PREFIX,
+        }
+    }
+
+    fn b58_script_address_prefix(&self) -> [u8; 2] {
+        match self {
+            Self::Consensus(params) => params.b58_script_address_prefix(),
+            Self::RegTest { .. } => constants::regtest::B58_SCRIPT_ADDRESS_PREFIX,
+        }
+    }
+}
+
+/// A decrypted note.
+struct DecryptedNote<D: Domain> {
+    /// The incoming viewing key used to decrypt the note.
+    ivk: D::IncomingViewingKey,
+    /// The recipient of the note.
+    recipient: D::Recipient,
+    /// The note!
+    note: D::Note,
+    /// The memo sent with the note.
+    memo: D::Memo,
+}
+
+impl<D: Domain> fmt::Debug for DecryptedNote<D>
+where
+    D::IncomingViewingKey: fmt::Debug,
+    D::Recipient: fmt::Debug,
+    D::Note: fmt::Debug,
+    D::Memo: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DecryptedNote")
+            .field("ivk", &self.ivk)
+            .field("recipient", &self.recipient)
+            .field("note", &self.note)
+            .field("memo", &self.memo)
+            .finish()
+    }
+}
+
+/// A value correlated with an output index.
+struct OutputIndex<V> {
+    /// The index of the output within the corresponding shielded bundle.
+    output_index: usize,
+    /// The value for the output index.
+    value: V,
+}
```

Comment:
We don't appear to use any other `V` than `Option<DecryptedNote<D>>`, so could this more simply be
```suggestion
/// An output index and possibly decrypted note.
struct Output<D> {
    /// The index of the output within the corresponding shielded bundle.
    output_index: usize,
    /// The decrypted note if available.
    note: Option<DecryptedNote<D>>,
}
```